# Notebook Intro:

In this notebook, I import the training data and perform basic clean (review nulls and datatypes). 

I then output the cleaned data to a csv so other notebooks can use the clean data to create a model.

Note that the data dictionary is [here](http://jse.amstat.org/v19n3/decock/DataDocumentation.txt).

## import packages and data

In [4]:
# imports
import numpy as np
import pandas as pd

In [5]:
# import training data
filepath = '../datasets/Original/train.csv'

train = pd.read_csv(filepath)
train.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley  \
0  109  533352170           60        RL           NaN     13517   Pave   NaN   
1  544  531379050           60        RL          43.0     11492   Pave   NaN   
2  153  535304180           20        RL          68.0      7922   Pave   NaN   
3  318  916386060           60        RL          73.0      9802   Pave   NaN   
4  255  906425045           50        RL          82.0     14235   Pave   NaN   

  Lot Shape Land Contour  ... Screen Porch Pool Area Pool QC Fence  \
0       IR1          Lvl  ...            0         0     NaN   NaN   
1       IR1          Lvl  ...            0         0     NaN   NaN   
2       Reg          Lvl  ...            0         0     NaN   NaN   
3       Reg          Lvl  ...            0         0     NaN   NaN   
4       IR1          Lvl  ...            0         0     NaN   NaN   

  Misc Feature Misc Val Mo Sold Yr Sold  Sale Type  SalePrice  
0          NaN        0       3    2010        WD      130500  
1          NaN        0       4    2009        WD      220000  
2          NaN        0       1    2010        WD      109000  
3          NaN        0       4    2010        WD      174000  
4          NaN        0       3    2010        WD      138500  

[5 rows x 81 columns]

## Check on general features of the data:

* review datatypes 
* review if any data is missing

In [6]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2051 entries, 0 to 2050
Data columns (total 81 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               2051 non-null   int64  
 1   PID              2051 non-null   int64  
 2   MS SubClass      2051 non-null   int64  
 3   MS Zoning        2051 non-null   object 
 4   Lot Frontage     1721 non-null   float64
 5   Lot Area         2051 non-null   int64  
 6   Street           2051 non-null   object 
 7   Alley            140 non-null    object 
 8   Lot Shape        2051 non-null   object 
 9   Land Contour     2051 non-null   object 
 10  Utilities        2051 non-null   object 
 11  Lot Config       2051 non-null   object 
 12  Land Slope       2051 non-null   object 
 13  Neighborhood     2051 non-null   object 
 14  Condition 1      2051 non-null   object 
 15  Condition 2      2051 non-null   object 
 16  Bldg Type        2051 non-null   object 
 17  House Style   

In [7]:
train['Mas Vnr Area'].value_counts()

0.0      1216
120.0      11
176.0      10
200.0      10
180.0       9
         ... 
57.0        1
572.0       1
65.0        1
567.0       1
161.0       1
Name: Mas Vnr Area, Length: 373, dtype: int64

In [8]:
train[train['Lot Frontage'].notna()]['Lot Frontage'].sort_values()
#based on results from this - the na could be a 0 sized lot rather than a missing value - it's not clear

53       21.0
1534     21.0
787      21.0
796      21.0
1834     21.0
        ...  
1263    174.0
441     195.0
1875    200.0
1335    313.0
960     313.0
Name: Lot Frontage, Length: 1721, dtype: float64

**Analyze the following columns with data that is a float datatype and some are NA**
* if the data has a 0 this means the NA value is likely an error and wouldn't correspond to 0

In [9]:
train['BsmtFin SF 1'].value_counts()

0.0      657
24.0      17
16.0      10
300.0      7
384.0      7
        ... 
848.0      1
70.0       1
688.0      1
472.0      1
661.0      1
Name: BsmtFin SF 1, Length: 821, dtype: int64

In [10]:
train['BsmtFin SF 2'].value_counts()

0.0      1803
294.0       5
162.0       3
539.0       3
72.0        3
         ... 
712.0       1
529.0       1
215.0       1
252.0       1
823.0       1
Name: BsmtFin SF 2, Length: 205, dtype: int64

In [11]:
train['Bsmt Unf SF'].value_counts()

0.0       165
384.0      14
672.0      11
728.0      10
816.0       9
         ... 
494.0       1
1323.0      1
1041.0      1
314.0       1
2062.0      1
Name: Bsmt Unf SF, Length: 967, dtype: int64

In [12]:
train['Total Bsmt SF'].value_counts()

0.0       54
864.0     51
672.0     20
1040.0    20
816.0     17
          ..
842.0      1
752.0      1
1232.0     1
1480.0     1
1125.0     1
Name: Total Bsmt SF, Length: 892, dtype: int64

In [13]:
train['Bsmt Full Bath'].value_counts()

0.0    1200
1.0     824
2.0      23
3.0       2
Name: Bsmt Full Bath, dtype: int64

In [14]:
train['Bsmt Half Bath'].value_counts()

0.0    1923
1.0     122
2.0       4
Name: Bsmt Half Bath, dtype: int64

In [15]:
train['Garage Cars'].value_counts()

2.0    1136
1.0     524
3.0     263
0.0     113
4.0      13
5.0       1
Name: Garage Cars, dtype: int64

In [16]:
train['Garage Area'].value_counts()

0.0       113
440.0      70
576.0      69
484.0      52
528.0      46
         ... 
406.0       1
309.0       1
580.0       1
1184.0      1
557.0       1
Name: Garage Area, Length: 514, dtype: int64

The following column, 'Garage Yr Blt' does not have zeroes, so I review to see if the NA's correspond to when the Garage type is NaN (which means that the property has no garage).  

The following two lines of code and analysis show that 'Garage Yr Blt' being NA corresponds to no garage being built. So I'll convert the NA's in the 'Garage Yr Blt' column to another value.  Since this is a float column, I'll convert the Na's to 0's and make a note of it in the file.

In [17]:
train[train['Garage Type'].isnull()]['Garage Yr Blt'].sort_values()

28     NaN
53     NaN
65     NaN
79     NaN
101    NaN
        ..
1991   NaN
2010   NaN
2027   NaN
2039   NaN
2042   NaN
Name: Garage Yr Blt, Length: 113, dtype: float64

In [18]:
train[train['Garage Type'].notna()]['Garage Yr Blt'].sort_values()

1793    1895.0
410     1896.0
217     1900.0
520     1900.0
12      1900.0
         ...  
589     2009.0
1241    2010.0
144     2010.0
1699    2207.0
1712       NaN
Name: Garage Yr Blt, Length: 1938, dtype: float64

### Clean null values
A few columns have almost no data, so look into what NA means:

|Column Name|  total non-null entries| approximate % of data| what NA means| convert/remove NA
| :-: | :-: | :-: | :- |:- |
|'Pool QC'| 9 | 0.4%| per dictionary, NA means No Pool, so keep this column (NA has meaning)|convert to none
|'Misc Feature'| 65 | 3.2%| per dictionary, NA means none, so can keep this column (NA has meaning)|convert to none
|'Alley'| 140 | 6.8%| per dictionary, NA means no alley access NA has meaning, so keep)|convert to none
|'Fence'| 400 | 9.5% |per dictionary, NA means No Fence, so keep this column (NA has meaning)|convert to none

**all columns with high number of NA values have meaning to the NA, so keeping these columns.**



A few columns have some entries missing, so look into what NA means for these:

|index| Column Name| non-null entries | data type in column| analysis on what the NA means| convert/remove NA entries|
| :-: | :-: | :-: | :-: | :-: |:-: |
| 4  | Lot Frontage   |  1721 non-null |  float64  |   based on analysis above, na could be values of 0| because this is a data type float category, convert to 0 with note in file that 0 means it was previously missing data
| 26 | Mas Vnr Type   |  2029 non-null |  object   |   per dictionary - none is entered if no masonry, so Na is an error| remove
| 27 | Mas Vnr Area   |  2029 non-null |  float64  |   per analysis - 0 is entered as an area, so NA is an error | remove
| 31 | Bsmt Qual      |  1996 non-null |  object   |   per dictionary - NA means no basement | convert to none
| 32 | Bsmt Cond      |  1996 non-null |  object   |   per dictionary - NA means no basement | convert to none
| 33 | Bsmt Exposure  |  1993 non-null |  object   |   per dictionary - NA means no basement | convert to none
| 34 | BsmtFin Type 1 |  1996 non-null |  object   |   per dictionary - NA means no basement | convert to none
| 35 | BsmtFin SF 1   |  2050 non-null |  float64  |   per analysis - 0 is entered as an area, so NA is an error | remove
| 36 | BsmtFin Type 2 |  1995 non-null |  object   |   per dictionary - NA means no basement | convert to none
| 37 | BsmtFin SF 2   |  2050 non-null |  float64  |   per analysis - 0 is entered as an area, so NA is an error | remove
| 38 | Bsmt Unf SF    |  2050 non-null |  float64  |   per analysis - 0 is entered as an area, so NA is an error | remove
| 39 | Total Bsmt SF  |  2050 non-null |  float64  |   per analysis - 0 is entered as an area, so NA is an error | remove
| 48 | Bsmt Full Bath |  2049 non-null |  float64  |   per analysis - 0 is entered as an area, so NA is an error | remove
| 49 | Bsmt Half Bath |  2049 non-null |  float64  |   per analysis - 0 is entered as an area, so NA is an error | remove
| 58 | Fireplace Qu   |  1051 non-null |  object   |   per dictionary - NA means no fireplace | convert to none
| 59 | Garage Type    |  1938 non-null |  object   |   per dictionary - NA means no garage   | convert to none
| 60 | Garage Yr Blt  |  1937 non-null |  float64  |   based on analysis - NA means no garage|because this is a data type float category, convert to 0 with note in file that 0 means it was previously missing data
| 61 | Garage Finish  |  1937 non-null |  object   |   per dictionary - NA means no garage   | convert to none
| 62 | Garage Cars    |  2050 non-null |  float64  |   per analysis - 0 is entered as an area, so NA is an error | remove 
| 63 | Garage Area    |  2050 non-null |  float64  |   per analysis - 0 is entered as an area, so NA is an error | remove
| 64 | Garage Qual    |  1937 non-null |  object   |   per dictionary - NA means no garage   | convert to none
| 65 | Garage Cond    |  1937 non-null |  object   |   per dictionary - NA means no garage   | convert to none
 


**Prior to removing null values that don't have meaning (which affects 30 to 40 rows total or < 2%), 
convert null values as appropriate based on review above**

In [19]:
# function to convert missing values to a selected value

def na_to_replacement_value(dataframe,list_of_column_names,replaced_value):
    for column in list_of_column_names:
        dataframe[column] = dataframe[column].replace(np.nan,replaced_value)          
    return dataframe

# perform function on the following columns and replace with none
# note that all columns have type object 

columns_na_to_none = ['Pool QC','Misc Feature','Alley','Fence','Bsmt Qual','Bsmt Cond','Bsmt Exposure','BsmtFin Type 1','BsmtFin Type 2','Fireplace Qu','Garage Type','Garage Finish','Garage Qual','Garage Cond']
train = na_to_replacement_value(train,columns_na_to_none,'none')

# convert the columns with missing values and float data types to 0

columns_na_to_0 = ['Lot Frontage','Garage Yr Blt']
train = na_to_replacement_value(train,columns_na_to_0,0)


train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2051 entries, 0 to 2050
Data columns (total 81 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               2051 non-null   int64  
 1   PID              2051 non-null   int64  
 2   MS SubClass      2051 non-null   int64  
 3   MS Zoning        2051 non-null   object 
 4   Lot Frontage     2051 non-null   float64
 5   Lot Area         2051 non-null   int64  
 6   Street           2051 non-null   object 
 7   Alley            2051 non-null   object 
 8   Lot Shape        2051 non-null   object 
 9   Land Contour     2051 non-null   object 
 10  Utilities        2051 non-null   object 
 11  Lot Config       2051 non-null   object 
 12  Land Slope       2051 non-null   object 
 13  Neighborhood     2051 non-null   object 
 14  Condition 1      2051 non-null   object 
 15  Condition 2      2051 non-null   object 
 16  Bldg Type        2051 non-null   object 
 17  House Style   

In [20]:
# after analysis remaining null values are likely error, so remove from dataframe

train_clean = train.dropna()

In [21]:
train_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2026 entries, 0 to 2050
Data columns (total 81 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               2026 non-null   int64  
 1   PID              2026 non-null   int64  
 2   MS SubClass      2026 non-null   int64  
 3   MS Zoning        2026 non-null   object 
 4   Lot Frontage     2026 non-null   float64
 5   Lot Area         2026 non-null   int64  
 6   Street           2026 non-null   object 
 7   Alley            2026 non-null   object 
 8   Lot Shape        2026 non-null   object 
 9   Land Contour     2026 non-null   object 
 10  Utilities        2026 non-null   object 
 11  Lot Config       2026 non-null   object 
 12  Land Slope       2026 non-null   object 
 13  Neighborhood     2026 non-null   object 
 14  Condition 1      2026 non-null   object 
 15  Condition 2      2026 non-null   object 
 16  Bldg Type        2026 non-null   object 
 17  House Style   

In [22]:
percent_data_lost = round((len(train)-len(train_clean))/len(train),4)*100
print(f'By eliminating the missing rows, the data frame went from {len(train)} to {len(train_clean)} rows, or it lost {percent_data_lost}% of the data')

By eliminating the missing rows, the data frame went from 2051 to 2026 rows, or it lost 1.22% of the data


This is a reasonable amount of data to lose. So this is fine.

In [24]:
# export clean data
filepath = '../datasets/interim_files/train_clean.csv'

train_clean.to_csv(filepath,index=False)


# Notebook Summary of Work:

In this notebook, I imported the training data and performed basic clean up (review nulls and datatypes). 

I then output the cleaned data to a csv so other notebooks can use the clean data to create a model.